rmf_task/src/rmf_task/Estimate.cpp

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include <unordered_map>
+#include <mutex>
+
+#include <rmf_task/Estimate.hpp>
+
+namespace rmf_task {
+
+//==============================================================================
+class Estimate::Implementation
+{
+public:
+
+  Implementation(agv::State finish_state, rmf_traffic::Time wait_until)
+  : _finish_state(std::move(finish_state)),
+    _wait_until(std::move(wait_until))
+  {}
+
+  agv::State _finish_state;
+  rmf_traffic::Time _wait_until;
+};
+
+//==============================================================================
+Estimate::Estimate(agv::State finish_state, rmf_traffic::Time wait_until)
+: _pimpl(rmf_utils::make_impl<Implementation>(
+    std::move(finish_state), std::move(wait_until)))
+{}
+
+//==============================================================================
+agv::State Estimate::finish_state() const
+{
+  return _pimpl->_finish_state;
+}
+
+//==============================================================================
+Estimate& Estimate::finish_state(agv::State new_finish_state)
+{
+  _pimpl->_finish_state = std::move(new_finish_state);
+  return *this;
+}
+
+//==============================================================================
+rmf_traffic::Time Estimate::wait_until() const
+{
+  return _pimpl->_wait_until;
+}
+
+//==============================================================================
+Estimate& Estimate::wait_until(rmf_traffic::Time new_wait_until)
+{
+  _pimpl->_wait_until = std::move(new_wait_until);
+  return *this;
+}
+
+//==============================================================================
+class EstimateCache::Implementation
+{
+public:
+
+  Implementation(std::size_t N)
+  : _cache(N, PairHash(N))
+  {
+
+  }
+
+  struct PairHash
+  {
+    PairHash(std::size_t N)
+    {
+      _shift = std::ceil(std::log2(N));
+    }
+
+    size_t operator()(const std::pair<size_t,size_t>& p) const
+    {
+      return p.first + (p.second << _shift);
+    }
+
+    std::size_t _shift;
+  };
+
+  using Cache = std::unordered_map<std::pair<size_t,size_t>,
+    CacheElement, PairHash>;
+
+  Cache _cache;
+  std::shared_ptr<std::mutex> _mutex = std::make_shared<std::mutex>();
```

Comment:
I think making this a `std::shared_ptr<std::mutex>` is solving the wrong problem.

My guess is that this choice was made because you were getting a compilation error from `rmf_utils::impl_ptr` about the copy constructor of `Implementation` being automatically deleted because the copy constructor of `std::mutex` is deleted. Switching to `std::shared_ptr<std::mutex>` seemed to magically make that compilation error disappear, and that's how we got here.

There's a problem here, though: This means every instance of `EstimateCache` that gets copied from one instance will share the same mutex, even though they're all storing data independently. The mutex will be "protecting" all of those instances from simultaneously read/writes even though the data of those instances is independent. This blocking will be completely unnecessary overhead.

Broadly speaking there are two options that I would recommend for making this all work correctly:

1. Use a `std::mutex` in `Implementation` and then use a `rmf_utils::unique_impl_ptr<Implementation>` for `EstimateCache` instead of `rmf_utils::impl_ptr<Implementation>`. This is the most straightforward solution if `EstimateCache` doesn't need to be copied.

2. Group together the `std::mutex` and `Cache` into their own struct and keep a `std::shared_ptr` of that whole struct. E.g.:

```
class EstimateCache::Implementation
{
  /* ... other stuff ... */

  struct Cache
  {
    using Data = std::unordered_map<std::pair<std::size_t, std::size_t>, CacheElement, PairHash>;
    Data data;
    std::mutex mutex;
  };

  std::shared_ptr<Cache> cache;
};
```

That way the mutex and the data are always tied together correctly. Copying may also be more efficient this way.

Reply:
You're absolutely right about the reason behind using a shared ptr for the mutex. Since `EstimateCache` does not need to be copied, I've adopted the first approach. Thank you for highlighting this problem.
https://github.com/osrf/rmf_core/pull/217/commits/89a68c0f0bbc8a26994c302c37f9d275713d141a